File: impl/sort/sort.py
import os
import random
import util

h_file = os.path.join(os.path.dirname(__file__), "sort.h")
c_file = os.path.join(os.path.dirname(__file__), "sort.c")
(ffi, lib) = util.compile("sort", h_file, c_file)

def c_bubblesort(arr):
    c_arr = ffi.new("int[]", arr)
    lib.bubblesort(c_arr, len(arr))
    result = ffi.unpack(c_arr, len(arr))
    return result


def c_insertionsort(arr):
    c_arr = ffi.new("int[]", arr)
    lib.insertionsort(c_arr, len(arr))
    result = ffi.unpack(c_arr, len(arr))
    return result


def c_mergesort(arr):
    c_arr = ffi.new("int[]", arr)
    lib.mergesort(c_arr, len(arr))
    result = ffi.unpack(c_arr, len(arr))
    return result


def c_qsort_builtin(arr):
    c_arr = ffi.new("int[]", arr)
    lib.qsort_builtin(c_arr, len(arr))
    result = ffi.unpack(c_arr, len(arr))
    return result


def c_quicksort(arr):
    c_arr = ffi.new("int[]", arr)
    lib.quicksort(c_arr, len(arr))
    result = ffi.unpack(c_arr, len(arr))
    return result


def py_builtin(arr):
    result = arr[:]
    result.sort()
    return result


def py_mergesort(arr):
    result = arr[:]
    mergesort(result)
    return result


def py_quicksort(arr):
    result = arr[:]
    quicksort(result)
    return result


def adapt_run_args(raw_args):
    # RUN: Echo back the specified values as integers in a list
    arr = [int(a) for a in raw_args]
    return {"arr": arr}


def adapt_benchmark_args(raw_args):
    # BENCHMARK: Generate and return a list of length N
    mode = raw_args[0]
    length = int(raw_args[1])

    if mode == "sorted":
        arr = list(range(1, length + 1))
    elif mode == "reversed":
        arr = list(range(length, 0, -1))
    elif mode == "zero":
        arr = [0] * length
    else:
        arr = random.sample(range(1, length + 1), length)

    return {"arr": arr}


def mergesort(arr):
    if len(arr) <= 1:
        return

    middle = len(arr) // 2
    h1 = arr[:middle]
    h2 = arr[middle:]

    mergesort(h1)
    mergesort(h2)

    idx1 = 0
    idx2 = 0

    for i in range(len(arr)):
        if idx1 < len(h1) and (idx2 == len(h2) or h1[idx1] < h2[idx2]):
            arr[i] = h1[idx1]
            idx1 += 1
        else:
            arr[i] = h2[idx2]
            idx2 += 1


def partition(arr, length, offset):
    pivot = arr[offset]
    i = offset + 1
    for j in range(offset + 1, offset + length):
        if arr[j] < pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[offset], arr[i-1] = arr[i-1], arr[offset]
    return i-1


def quicksort_internal(arr, length, offset):
    if length <= 1:
        return

    # Select the pivot at random to produce a more consistent result
    pivot_idx = random.randint(offset, offset + length - 1)
    arr[offset], arr[pivot_idx] = arr[pivot_idx], arr[offset]

    divider = partition(arr, length, offset)

    # Segment left of the pivot
    l_len = divider - offset
    l_offset = offset

    # Segment right of the pivot
    r_len = length - divider + offset - 1
    r_offset = divider + 1

    quicksort_internal(arr, l_len, l_offset)
    quicksort_internal(arr, r_len, r_offset)


def quicksort(arr):
    quicksort_internal(arr, len(arr), 0)

File: impl/sort/sort.c
#include <stdbool.h>
#include <stdlib.h>

void mergesort_internal(int *src, int *dest, int len)
{
    if (len <= 1) {
        return;
    }

    int mid = len / 2;
    int *src1 = src;
    int *src2 = src + mid;
    int *dest1 = dest;
    int *dest2 = dest + mid;
    int len1 = mid;
    int len2 = len - mid;

    mergesort_internal(dest1, src1, len1);
    mergesort_internal(dest2, src2, len2);

    int i = 0;
    int idx1 = 0;
    int idx2 = 0;
    while (idx1 < len1 || idx2 < len2) {
        if (idx2 == len2) {
            while (idx1 < len1) {
                dest[i++] = src1[idx1++];
            }
        }
        else if (idx1 == len1) {
            while (idx2 < len2) {
                dest[i++] = src2[idx2++];
            }
        }
        else if (src1[idx1] < src2[idx2]) {
            dest[i++] = src1[idx1++];
        }
        else {
            dest[i++] = src2[idx2++];
        }
    }
}

void bubblesort(int *arr, int len)
{
    for (int i = 0; i < len; i++)
    {
        bool swapped = false;

        for (int j = 0; j < len - 1; j++)
        {
            if (arr[j] > arr[j+1])
            {
                int temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
                swapped = true;
            }
        }

        if (!swapped)
        {
            // Short-circuit the operation if no swaps were performed. This means we ended up
            // with a sorted list (potentially) early.
            break;
        }
    }
}

void insertionsort(int *arr, int len)
{
    for (int i = 1; i < len; i++)
    {
        int curr = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > curr)
        {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = curr;
    }
}

void mergesort(int *arr, int len)
{
    int *result = malloc(len * sizeof(int));
    memcpy(result, arr, len * sizeof(int));
    mergesort_internal(arr, result, len);
    memcpy(arr, result, len * sizeof(int));
    free(result);
}

int cmpfunc (const void *a, const void *b)
{
   return *(int *)a - *(int *)b;
}

void qsort_builtin(int *arr, int len)
{
    qsort(arr, len, sizeof(int), cmpfunc);
}

void swap(int *x, int *y)
{
    int tmp = *x;
    *x = *y;
    *y = tmp;
}

int partition(int *arr, int len)
{
    int pivot = arr[0];
    int i = 1;
    for (int j = 1; j < len; j++) {
        if (arr[j] < pivot) {
            swap(&arr[j], &arr[i]);
            i++;
        }
    }

    swap(&arr[0], &arr[i-1]);
    return i-1;
}

void quicksort(int *arr, int len)
{
    if (len <= 1) {
        return;
    }

    // Select the pivot at random to produce a more consistent result
    int pivot_idx = rand() % len;
    swap(&arr[0], &arr[pivot_idx]);

    int divider = partition(arr, len);
    quicksort(arr, divider);
    quicksort(arr + divider + 1, len - divider - 1);
}